torchao/prototype/low_bit_optim/cpu_offload.py

```diff
@@ -0,0 +1,47 @@
+import torch
+from torch.optim.optimizer import Optimizer
+
+
+class CPUOffloadOptimizer:
+    def __init__(self, base_optimizer: Optimizer) -> None:
+        self.optim = base_optimizer
+        self.param_cuda2cpu_map = dict()
+        self.d2h_grad_stream = torch.cuda.Stream()
+
+        def copy_grad_hook(p_cuda):
+            if p_cuda.grad is not None:
+                p_cpu = self.param_cuda2cpu_map[p_cuda]
```

Comment:
I think we need a `self.d2h_grad_stream.wait_stream(torch.cuda.current_stream())` or else the D2H copy may not see the correct values in `p_cuda.grad`. This should be why your backward kernels are finishing after your D2H copies.

It is interesting that loss looks good 😃

Reply:
I'm testing with ViT fine-tuning, so I guess it's quite forgiving to bugs 🤣